fdbserver/TagPartitionedLogSystem.actor.cpp

```diff
@@ -117,12 +217,47 @@ struct TagPartitionedLogSystem : ILogSystem, ReferenceCounted<TagPartitionedLogS
 		return dbgid;
 	}
 
+	void addPseudoLocality(int8_t locality) override {
+		ASSERT(locality < 0);
+		pseudoLocalities.insert(locality);
+		pseudoLocalityPopVersion[locality] = 0;
+	}
+
+	Tag getPseudoPopTag(Tag tag, ProcessClass::ClassType type) override {
+		switch (type) {
+		case ProcessClass::LogRouterClass:
+			if (tag.locality == tagLocalityLogRouter && pseudoLocalities.count(tag.locality) > 0) {
+				tag.locality = tagLocalityLogRouterMapped;
+			}
+			break;
+
+		default:
+			break;
+		}
+		return tag;
+	}
+
+	bool isPseudoLocality(int8_t locality) override {
+		return pseudoLocalities.count(locality) > 0;
+	}
+
+	Version getPseudoLocalityPopVersion(int8_t locality, Version upTo) override {
```

Comment:
this function is modifying state (the pseudoLocalityPopVersion map) so it should be renamed as popPseudoLocalityTag